README.md
<h1 align="center">
  Flutter Multi Image Picker
  <br>
</h1>

Original plugin by Sh1d0w, forked by Guzram

<h4 align="center">
  <a href="https://flutter.io" target="_blank">Flutter</a> plugin that allows you to display multi image picker on iOS and Android.
</h4>

<p align="center">
  <a href="https://pub.dartlang.org/packages/multi_image_picker">
    <img src="https://img.shields.io/travis/Sh1d0w/multi_image_picker.svg"
         alt="Build">
  </a>
  <a href="https://pub.dartlang.org/packages/multi_image_picker"><img src="https://img.shields.io/pub/v/multi_image_picker.svg"></a>
  <a href="https://www.buymeacoffee.com/Sh1d0w">
    <img src="https://img.shields.io/badge/$-donate-ff69b4.svg?maxAge=2592000&amp;style=flat">
  </a>
</p>

<p align="center">
  <a href="#key-features">Key Features</a> •
  <a href="https://sh1d0w.github.io/multi_image_picker/#/">Documentation</a> •
  <a href="https://sh1d0w.github.io/multi_image_picker/#/faq">FAQ</a> •
  <a href="#credits">Credits</a> •
  <a href="#related">Related</a> •
  <a href="#license">License</a>
</p>

<p align="center">
  <img src="https://github.com/idabgsram/multi_image_picker/raw/master/screenshots/hero.png" alt="Multi Image Picker" />
</p>

## Key Features

- Pick multiple images
- Native performance
- Photos sorted by albums
- Take a picture option in the grid view
- Restrict the maximum count of images the user can pick
- Customizable UI and localizations
- Thumbnail support
- Specify the image quality of the original image or thumbnails
- Read image meta data

## Documentation

You can see the official documentation [here](https://sh1d0w.github.io/multi_image_picker/#/).

## Upgrading

For more information regarding breaking changes and migrating through different versions of the plugin please read the [upgrading guide](https://sh1d0w.github.io/multi_image_picker/#/upgrading).

## Credits

This software uses the following open source packages:

- [BSImagePicker](https://github.com/mikaoj/BSImagePicker) - iOS
- [FishBun](https://github.com/sangcomz/FishBun) - Android

## Related

[image_picker](https://pub.dartlang.org/packages/image_picker) - Official Flutter image picker plugin

## Support

If this plugin was useful to you, helped you to deliver your app, saved you a lot of time, or you just want to support the project, I would be very grateful if you buy me a cup of coffee.

<a href="https://www.buymeacoffee.com/Sh1d0w" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/purple_img.png" alt="Buy Me A Coffee" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;" ></a>

## License

MIT


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/)
and this project adheres to [Semantic Versioning](https://semver.org/).

### Change (v4.8.00)

## 2021-03-06

- Add Null safery and bumb the minimum Flutter version required.

### Change (v4.7.15)

## 2021-02-09

- Bump glide version on Android

### Change (v4.7.14)

## 2020-09-16

- Revert a commit, causing a bug with mage selection on iOS

### Change (v4.7.13)

## 2020-09-09

- Fix picking live images on iOS will now pick the correct image

### Change (v4.7.12)

## 2020-08-22

- README and CHANGELOG fixes for pub.dev

### Change (v4.7.11)

## 2020-08-19

- Fix image picker regression on Android

### Change (v4.7.10)

## 2020-07-31

- Support activity in v1 embedding (https://github.com/Sh1d0w/multi_image_picker/pull/506)
- Fix on iOS can't present picker in some cases (https://github.com/Sh1d0w/multi_image_picker/pull/507)

### Change (v4.6.9)

## 2020-07-22

- Downgrade minimum Flutter version

### Change (v4.6.8)

## 2020-07-21

- Supporting the new Android plugins APIs

### Change (v4.6.7)

## 2020-04-30

- Revert fix resizing an image causes jagged lines on Android

### Change (v4.6.6)

## 2020-04-11

- Fix resizing an image causes jagged lines on Android

### Change (v4.6.5+rc.1)

## 2020-03-27

- Disable iOS 13 swipe to dismiss - to force user to press cancel or done.

### Change (v4.6.4)

## 2020-03-21

- Fix quality parameter on iOS that always was evaluated to 0

### Change (v4.6.4-rc.1)

## 2020-03-21

- Add support for Android minSdkVersion 16 and up & fixing incorrect rotation in Android 10 [#414](https://github.com/Sh1d0w/multi_image_picker/pull/414)

### Change (v4.6.3)

## 2020-03-07

- Revert Adding support for Android API Level 16 and up [#339](https://github.com/Sh1d0w/multi_image_picker/pull/399)

### Change (v4.6.2)

## 2020-03-05

- Adding support for Android API Level 16 and up [#339](https://github.com/Sh1d0w/multi_image_picker/pull/399)

### Change (v4.6.1)

## 2020-01-04

- Update FishBun version and remove permission handling from the library. FishBun and BSImagePicker handle permission requests by themselves

### Change (v4.6.0-rc.3)

## 2019-11-21

- Remove direct access to file URL, beacause of breaking change in iOS 13. Please use `getByteData` or `getThumbData` to obtain the images.

### Change (v4.6.0-rc.2)

## 2019-11-20

- Annotate controller type with UIViewController [#316](https://github.com/Sh1d0w/multi_image_picker/pull/316)

### Change (v4.6.0-rc.1)

## 2019-11-08

- BREAKING CHANGE: Compatibility for [breaking change in Flutter 1.10.15](https://groups.google.com/forum/#!topic/flutter-announce/lUKzLAd8OG8)

### Changed (v4.5.9)

## 2019-10-29

- Images that fail to download from iCloud on iOS will now throw `AssetFailedToDownloadException` exception if you try to access the path.

### Changed (v4.5.8)

## 2019-10-09

- `getByteData` and `getThumbByteData` now will return the current image with all adjustments on iOS [#277](https://github.com/Sh1d0w/multi_image_picker/issues/277)

### Changed (v4.5.7)

## 2019-10-06

- Add new option for both Android and iOS `autoCloseOnSelectionLimit`. It will close the image picker as soon as the selection limit is reached.

### Changed (v4.5.6)

## 2019-09-26

- Fix crash on iOS when optimize storage is enabled

### Changed (v4.5.5)

## 2019-09-16

- Update BSImagePicker version to 2.10.2

### Changed (v4.5.4)

## 2019-09-11

- Update BSImagePicker version to 2.10.1

### Changed (v4.5.3)

## 2019-09-03

- Update FishBun to version 0.11.1
- Fixed picking image using camera does not return FilePath [#226](https://github.com/Sh1d0w/multi_image_picker/issues/226)

### Changed (v4.5.2)

## 2019-08-19

- Trying to read images that does not exists will now throw AssetNotFoundException [#222](https://github.com/Sh1d0w/multi_image_picker/issues/222)

### Changed (v4.5.1)

## 2019-08-17

- Permission Denied error now will be correctly thrown on iOS if the user has disabled camera access.

### Changed (v4.5.0+2)

## 2019-08-13

- Don't pin meta package version

### Changed (v4.5.0+1)

## 2019-08-13

- Make thumb provider use non depracated plugin methods
- Temporarily pin meta package version to 1.1.7

### Changed (v4.5.0)

## 2019-08-12

- The plugin now returns file paths as well. To obtain the asset file path use `await asset.filePath`
- Removed `deleteImages` method. This plugin purpose is only to pick images.
- Deprecated `Asset.requestThumbnail` method. Use `Asset.getThumbByteData` instead.
- Deprecated `Asset.requestOriginal` method. Use `Asset.getByteData` instead.
- Deprecated `Asset.requestMetadata` method. Use `Asset.metadata` instead.

### Changed (v4.4.1)

## 2019-08-06

- Pin minimum Flutter version required to 1.7.8

### Changed (v4.4.0)

## 2019-08-06

- Added Material option `textOnNothingSelected` [#201](https://github.com/Sh1d0w/multi_image_picker/issues/201)
- Updated the code to use defaultBinaryMessenger instead of the deprecated BinaryMessenger
- `pickImages` now throws NoImagesSelectedException, PermissionDeniedException and PermissionPermanentlyDeniedExeption in addition to PlatformException. You can implement different logic for handling each case separately now.
- dateTime, dateTimeOriginal and dateTimeDigitized in the Metadata.exif object will now properly be returned as strings. The format is "YYYY:MM:DD HH:MM:SS" with time shown in 24-hour format as per Exif spec.
- Added the ability to change the icons for back and done buttons on Android. For more information see the [documentation](https://sh1d0w.github.io/multi_image_picker/#/theming) and the example app in this repository.

### Changed (v4.3.6)

## 2019-07-18

- Added new Android option useDetailsView to disable details view when clicking on the image and directly select it

### Changed (v4.3.5)

## 2019-07-07

- Update FishBun library to version 0.10.0

### Changed (v4.3.4)


test/multi_image_picker_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:multi_image_picker/multi_image_picker.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  group('MultiImagePicker', () {
    const MethodChannel channel = MethodChannel('multi_image_picker');

    final List<MethodCall> log = <MethodCall>[];

    setUp(() {
      channel.setMockMethodCallHandler((MethodCall methodCall) async {
        log.add(methodCall);
        if (methodCall.method == 'requestOriginal' ||
            methodCall.method == 'requestThumbnail') {
          return true;
        }
        return [
          {'identifier': 'SOME_ID_1'},
          {'identifier': 'SOME_ID_2'}
        ];
      });

      log.clear();
    });

    group('#pickImages', () {
      test('passes max images argument correctly', () async {
        await MultiImagePicker.pickImages(maxImages: 5);

        expect(
          log,
          <Matcher>[
            isMethodCall('pickImages', arguments: <String, dynamic>{
              'maxImages': 5,
              'enableCamera': false,
              'iosOptions': CupertinoOptions().toJson(),
              'androidOptions': MaterialOptions().toJson(),
              'selectedAssets': [],
            }),
          ],
        );
      });

      test('passes selected assets correctly', () async {
        Asset asset = Asset("test", "test.jpg", 100, 100);
        await MultiImagePicker.pickImages(
          maxImages: 5,
          selectedAssets: [asset],
        );

        expect(
          log,
          <Matcher>[
            isMethodCall('pickImages', arguments: <String, dynamic>{
              'maxImages': 5,
              'enableCamera': false,
              'iosOptions': CupertinoOptions().toJson(),
              'androidOptions': MaterialOptions().toJson(),
              'selectedAssets': [asset.identifier],
            }),
          ],
        );
      });

      test('passes cuppertino options argument correctly', () async {
        CupertinoOptions cupertinoOptions = CupertinoOptions(
          backgroundColor: '#ffde05',
          selectionCharacter: 'A',
          selectionFillColor: '#004ed5',
          selectionShadowColor: '#05e43d',
          selectionStrokeColor: '#0f5e4D',
          selectionTextColor: '#ffffff',
        );

        await MultiImagePicker.pickImages(
            maxImages: 5, cupertinoOptions: cupertinoOptions);

        expect(
          log,
          <Matcher>[
            isMethodCall('pickImages', arguments: <String, dynamic>{
              'maxImages': 5,
              'enableCamera': false,
              'iosOptions': cupertinoOptions.toJson(),
              'androidOptions': MaterialOptions().toJson(),
              'selectedAssets': [],
            }),
          ],
        );
      });

      test('passes meterial options argument correctly', () async {
        MaterialOptions materialOptions = MaterialOptions(
          actionBarTitle: "Aciton bar",
          allViewTitle: "All view title",
          actionBarColor: "#aaaaaa",
          actionBarTitleColor: "#bbbbbb",
          lightStatusBar: false,
          statusBarColor: '#abcdef',
          startInAllView: true,
          useDetailsView: true,
          selectCircleStrokeColor: "#ffffff",
        );
        await MultiImagePicker.pickImages(
            maxImages: 5, materialOptions: materialOptions);

        expect(
          log,
          <Matcher>[
            isMethodCall('pickImages', arguments: <String, dynamic>{
              'maxImages': 5,
              'enableCamera': false,
              'androidOptions': materialOptions.toJson(),
              'iosOptions': CupertinoOptions().toJson(),
              'selectedAssets': [],
            }),
          ],
        );
      });

      test('does not accept a negative images count', () {
        expect(
          MultiImagePicker.pickImages(maxImages: -10),
          throwsArgumentError,
        );
      });

      test('throws correct exception on cancel', () {
        channel.setMockMethodCallHandler((MethodCall methodCall) async {
          throw PlatformException(code: 'CANCELLED', message: 'Some error');
        });

        expect(
          () => MultiImagePicker.pickImages(maxImages: 10),
          throwsA(isA<NoImagesSelectedException>()),
        );
      });

      test('throws correct exception when permission denied', () {
        channel.setMockMethodCallHandler((MethodCall methodCall) async {
          throw PlatformException(
              code: 'PERMISSION_DENIED', message: 'Some error');
        });

        expect(
          () => MultiImagePicker.pickImages(maxImages: 10),
          throwsA(isA<PermissionDeniedException>()),
        );
      });

      test('throws correct exception when permission permanently denied', () {
        channel.setMockMethodCallHandler((MethodCall methodCall) async {
          throw PlatformException(
              code: 'PERMISSION_PERMANENTLY_DENIED', message: 'Some error');
        });

        expect(
          () => MultiImagePicker.pickImages(maxImages: 10),
          throwsA(isA<PermissionPermanentlyDeniedExeption>()),
        );
      });
    });

    test('requestOriginal accepts correct params', () async {
      const String id = 'SOME_ID';
      const int quality = 100;
      await MultiImagePicker.requestOriginal(id, quality);

      expect(
        log,
        <Matcher>[
          isMethodCall('requestOriginal', arguments: <String, dynamic>{
            'identifier': id,
            'quality': quality,
          }),
        ],
      );
    });

    group('#requestThumbnail', () {
      const String id = 'SOME_ID';
      const int width = 100;
      const int height = 200;
      const int quality = 100;
      test('accepts correct params', () async {
        await MultiImagePicker.requestThumbnail(id, width, height, quality);

        expect(
          log,
          <Matcher>[
            isMethodCall('requestThumbnail', arguments: <String, dynamic>{
              'identifier': id,
              'width': width,
              'height': height,
              'quality': quality,
            }),
          ],
        );
      });

      test('does not accept a negative width or height', () {
        expect(
          MultiImagePicker.requestThumbnail(id, -100, height, quality),
          throwsArgumentError,
        );

        expect(
          MultiImagePicker.requestThumbnail(id, width, -100, quality),
          throwsArgumentError,
        );
      });
      test('does not accept invalid quality', () {
        expect(
          MultiImagePicker.requestThumbnail(id, -width, height, -100),
          throwsArgumentError,
        );

        expect(
          MultiImagePicker.requestThumbnail(id, width, height, 200),
          throwsArgumentError,
        );
      });
    });
  });
}


lib/src/asset.dart
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:multi_image_picker/multi_image_picker.dart';

class Asset {
  /// The resource identifier
  String? _identifier;

  /// The resource file name
  String? _name;

  /// Original image width
  int? _originalWidth;

  /// Original image height
  int? _originalHeight;

  Asset(
    this._identifier,
    this._name,
    this._originalWidth,
    this._originalHeight,
  );

  /// The BinaryChannel name this asset is listening on.
  String get _channel {
    return 'multi_image_picker/image/$_identifier';
  }

  String get _thumbChannel => '$_channel.thumb';

  String get _originalChannel => '$_channel.original';

  /// Returns the original image width
  int? get originalWidth {
    return _originalWidth;
  }

  /// Returns the original image height
  int? get originalHeight {
    return _originalHeight;
  }

  /// Returns true if the image is landscape
  bool get isLandscape {
    return _originalWidth! > _originalHeight!;
  }

  /// Returns true if the image is Portrait
  bool get isPortrait {
    return _originalWidth! < _originalHeight!;
  }

  /// Returns the image identifier
  String? get identifier {
    return _identifier;
  }

  /// Returns the image name
  String? get name {
    return _name;
  }

  /// Requests a thumbnail for the [Asset] with give [width] and [hegiht].
  ///
  /// The method returns a Future with the [ByteData] for the thumb,
  /// as well as storing it in the _thumbData property which can be requested
  /// later again, without need to call this method again.
  ///
  /// You can also pass the optional parameter [quality] to reduce the quality
  /// and the size of the returned image if needed. The value should be between
  /// 0 and 100. By default it set to 100 (max quality).
  ///
  /// Once you don't need this thumb data it is a good practice to release it,
  /// by calling releaseThumb() method.
  Future<ByteData> getThumbByteData(int width, int height,
      {int quality = 100}) async {
    if (width < 0) {
      throw new ArgumentError.value(width, 'width cannot be negative');
    }

    if (height < 0) {
      throw new ArgumentError.value(height, 'height cannot be negative');
    }

    if (quality < 0 || quality > 100) {
      throw new ArgumentError.value(
          quality, 'quality should be in range 0-100');
    }

    Completer completer = new Completer<ByteData>();
    ServicesBinding.instance.defaultBinaryMessenger
        .setMessageHandler(_thumbChannel, (ByteData? message) async {
      completer.complete(message);
      ServicesBinding.instance.defaultBinaryMessenger
          .setMessageHandler(_thumbChannel, null);
      return message;
    });

    await MultiImagePicker.requestThumbnail(
        _identifier, width, height, quality);
    return completer.future as FutureOr<ByteData>;
  }

  /// Requests the original image for that asset.
  ///
  /// You can also pass the optional parameter [quality] to reduce the quality
  /// and the size of the returned image if needed. The value should be between
  /// 0 and 100. By default it set to 100 (max quality).
  ///
  /// The method returns a Future with the [ByteData] for the image,
  /// as well as storing it in the _imageData property which can be requested
  /// later again, without need to call this method again.
  Future<ByteData> getByteData({int quality = 100}) async {
    if (quality < 0 || quality > 100) {
      throw new ArgumentError.value(
          quality, 'quality should be in range 0-100');
    }

    Completer completer = new Completer<ByteData>();
    ServicesBinding.instance.defaultBinaryMessenger
        .setMessageHandler(_originalChannel, (ByteData? message) async {
      completer.complete(message);
      ServicesBinding.instance.defaultBinaryMessenger
          .setMessageHandler(_originalChannel, null);
      return message;
    });

    await MultiImagePicker.requestOriginal(_identifier, quality);
    return completer.future as FutureOr<ByteData>;
  }

  /// Requests the original image meta data
  Future<Metadata> get metadata {
    return MultiImagePicker.requestMetadata(_identifier);
  }

  @Deprecated(
    'This method will be deprecated in the next major release. Please use getByteData method instead.',
  )
  Future<ByteData> requestOriginal({int quality = 100}) {
    return getByteData(quality: quality);
  }

  @Deprecated(
    'This method will be deprecated in the next major release. Please use getThumbByteData method instead.',
  )
  Future<ByteData> requestThumbnail(
    int width,
    int height, {
    int quality = 100,
  }) async {
    return getThumbByteData(width, height, quality: quality);
  }

  @Deprecated(
    'This method will be deprecated in the next major release. Please use metadata getter instead.',
  )
  Future<Metadata> requestMetadata() {
    return metadata;
  }
}


lib/src/asset_thumb.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:multi_image_picker/src/asset.dart';

class AssetThumb extends StatefulWidget {
  /// The asset we want to show thumb for.
  final Asset asset;

  /// The thumb width
  final int width;

  /// The thumb height
  final int height;

  /// The thumb quality
  final int quality;

  /// This is the widget that will be displayed while the
  /// thumb is loading.
  final Widget spinner;

  const AssetThumb({
    Key? key,
    required this.asset,
    required this.width,
    required this.height,
    this.quality = 100,
    this.spinner = const Center(
      child: SizedBox(
        width: 50,
        height: 50,
        child: CircularProgressIndicator(),
      ),
    ),
  }) : super(key: key);

  @override
  _AssetThumbState createState() => _AssetThumbState();
}

class _AssetThumbState extends State<AssetThumb> {
  ByteData? _thumbData;

  int get width => widget.width;
  int get height => widget.height;
  int get quality => widget.quality;
  Asset get asset => widget.asset;
  Widget get spinner => widget.spinner;

  @override
  void initState() {
    super.initState();
    this._loadThumb();
  }

  @override
  void didUpdateWidget(AssetThumb oldWidget) {
    if (oldWidget.asset.identifier != widget.asset.identifier) {
      this._loadThumb();
    }
    super.didUpdateWidget(oldWidget);
  }

  void _loadThumb() async {
    setState(() {
      _thumbData = null;
    });

    ByteData thumbData = await asset.getThumbByteData(
      width,
      height,
      quality: quality,
    );

    if (this.mounted) {
      setState(() {
        _thumbData = thumbData;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    if (_thumbData == null) {
      return spinner;
    }
    return Image.memory(
      _thumbData!.buffer.asUint8List(),
      key: ValueKey(asset.identifier),
      fit: BoxFit.cover,
      gaplessPlayback: true,
    );
  }
}


ios/Classes/SwiftMultiImagePickerPlugin.swift
import Flutter
import UIKit
import Photos
import BSImagePicker

extension PHAsset {
    
    var originalFilename: String? {
        
        var fname:String?
        
        if #available(iOS 9.0, *) {
            let resources = PHAssetResource.assetResources(for: self)
            if let resource = resources.last {
                fname = resource.originalFilename
            }
        }
        
        if fname == nil {
            // this is an undocumented workaround that works as of iOS 9.1
            fname = self.value(forKey: "filename") as? String
        }
        
        return fname
    }
}

fileprivate extension UIViewController {
    class func topViewController(base: UIViewController? = UIApplication.shared.keyWindow?.rootViewController) -> UIViewController? {
        if let presented = base?.presentedViewController {
            return topViewController(base: presented)
        }
        return base
    }
}

public class SwiftMultiImagePickerPlugin: NSObject, FlutterPlugin {
    var imagesResult: FlutterResult?
    var messenger: FlutterBinaryMessenger;

    let genericError = "500"

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger;
        super.init();
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "multi_image_picker", binaryMessenger: registrar.messenger())

        let instance = SwiftMultiImagePickerPlugin.init(messenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch (call.method) {
        case "pickImages":
            let status: PHAuthorizationStatus = PHPhotoLibrary.authorizationStatus()
            
            if (status == PHAuthorizationStatus.denied) {
                return result(FlutterError(code: "PERMISSION_PERMANENTLY_DENIED", message: "The user has denied the gallery access.", details: nil))
            }
            
            let vc = BSImagePickerViewController()
            
            if #available(iOS 13.0, *) {
                // Disables iOS 13 swipe to dismiss - to force user to press cancel or done.
                vc.isModalInPresentation = true
            }
            let arguments = call.arguments as! Dictionary<String, AnyObject>
            let maxImages = arguments["maxImages"] as! Int
            let enableCamera = arguments["enableCamera"] as! Bool
            let options = arguments["iosOptions"] as! Dictionary<String, String>
            let selectedAssets = arguments["selectedAssets"] as! Array<String>
            var totalImagesSelected = 0
            
            vc.maxNumberOfSelections = maxImages

            if (enableCamera) {
                vc.takePhotos = true
            }
            
            if selectedAssets.count > 0 {

                let options = PHFetchOptions()
                 if #available(iOS 9.1, *) {
                    let imagesPredicate = NSPredicate(format: "mediaType = %d", PHAssetMediaType.image.rawValue)
                    let liveImagesPredicate = NSPredicate(format: "(mediaSubtype & %d) != 0", PHAssetMediaSubtype.photoLive.rawValue)
                    let compound = NSCompoundPredicate(orPredicateWithSubpredicates: [imagesPredicate, liveImagesPredicate])
                    options.predicate = compound
                }

                let assets: PHFetchResult = PHAsset.fetchAssets(withLocalIdentifiers: selectedAssets, options: options)
                
                vc.defaultSelections = assets
            }

            if let takePhotoIcon = options["takePhotoIcon"] {
                if (!takePhotoIcon.isEmpty) {
                    vc.takePhotoIcon = UIImage(named: takePhotoIcon)
                }
            }

            if let backgroundColor = options["backgroundColor"] {
                if (!backgroundColor.isEmpty) {
                    vc.backgroundColor = hexStringToUIColor(hex: backgroundColor)
                }
            }

            if let selectionFillColor = options["selectionFillColor"] {
                if (!selectionFillColor.isEmpty) {
                    vc.selectionFillColor = hexStringToUIColor(hex: selectionFillColor)
                }
            }

            if let selectionShadowColor = options["selectionShadowColor"] {
                if (!selectionShadowColor.isEmpty) {
                    vc.selectionShadowColor = hexStringToUIColor(hex: selectionShadowColor)
                }
            }

            if let selectionStrokeColor = options["selectionStrokeColor"] {
                if (!selectionStrokeColor.isEmpty) {
                    vc.selectionStrokeColor = hexStringToUIColor(hex: selectionStrokeColor)
                }
            }

            if let selectionTextColor = options["selectionTextColor"] {
                if (!selectionTextColor.isEmpty) {
                    vc.selectionTextAttributes[NSAttributedString.Key.foregroundColor] = hexStringToUIColor(hex: selectionTextColor)
                }
            }

            if let selectionCharacter = options["selectionCharacter"] {
                if (!selectionCharacter.isEmpty) {
                    vc.selectionCharacter = Character(selectionCharacter)
                }
            }

            UIViewController.topViewController()?.bs_presentImagePickerController(vc, animated: true,
                select: { [weak vc](asset: PHAsset) -> Void in
                    totalImagesSelected += 1
                    

                    if let autoCloseOnSelectionLimit = options["autoCloseOnSelectionLimit"] {
                        if (!autoCloseOnSelectionLimit.isEmpty && autoCloseOnSelectionLimit == "true") {
                            if (maxImages == totalImagesSelected) {
                                guard let wVC = vc else {
                                    return
                                }
                                UIApplication.shared.sendAction(wVC.doneButton.action!, to: wVC.doneButton.target, from: self, for: nil)
                            }
                        }
                    }
                }, deselect: { (asset: PHAsset) -> Void in
                    totalImagesSelected -= 1
                }, cancel: { (assets: [PHAsset]) -> Void in
                    result(FlutterError(code: "CANCELLED", message: "The user has cancelled the selection", details: nil))
                }, finish: { (assets: [PHAsset]) -> Void in
                    var results = [NSDictionary]();
                    for asset in assets {
                        results.append([
                            "identifier": asset.localIdentifier,
                            "width": asset.pixelWidth,
                            "height": asset.pixelHeight,
                            "name": asset.originalFilename!
                        ]);
                    }
                    result(results);
                }, completion: nil)
            break;
        case "requestThumbnail":
            let arguments = call.arguments as! Dictionary<String, AnyObject>
            let identifier = arguments["identifier"] as! String
            let width = arguments["width"] as! Int
            let height = arguments["height"] as! Int
            let quality = arguments["quality"] as! Int
            let compressionQuality = Float(quality) / Float(100)
            let manager = PHImageManager.default()
            let options = PHImageRequestOptions()

            options.deliveryMode = PHImageRequestOptionsDeliveryMode.highQualityFormat
            options.resizeMode = PHImageRequestOptionsResizeMode.exact
            options.isSynchronous = false
            options.isNetworkAccessAllowed = true
            options.version = .current

            let assets: PHFetchResult = PHAsset.fetchAssets(withLocalIdentifiers: [identifier], options: nil)

            if (assets.count > 0) {
                let asset: PHAsset = assets[0];

                let ID: PHImageRequestID = manager.requestImage(
                    for: asset,
                    targetSize: CGSize(width: width, height: height),
                    contentMode: PHImageContentMode.aspectFill,
                    options: options,
                    resultHandler: {
                        (image: UIImage?, info) in
                        self.messenger.send(onChannel: "multi_image_picker/image/" + identifier + ".thumb", message: image?.jpegData(compressionQuality: CGFloat(compressionQuality)))
                        })

                if(PHInvalidImageRequestID != ID) {
                    return result(true);
                }
            }
            
            return result(FlutterError(code: "ASSET_DOES_NOT_EXIST", message: "The requested image does not exist.", details: nil))
        case "requestOriginal":
            let arguments = call.arguments as! Dictionary<String, AnyObject>
            let identifier = arguments["identifier"] as! String
            let quality = arguments["quality"] as! Int
            let compressionQuality = Float(quality) / Float(100)
            let manager = PHImageManager.default()
            let options = PHImageRequestOptions()

            options.deliveryMode = PHImageRequestOptionsDeliveryMode.highQualityFormat
            options.isSynchronous = false
            options.isNetworkAccessAllowed = true
            options.version = .current

            let assets: PHFetchResult = PHAsset.fetchAssets(withLocalIdentifiers: [identifier], options: nil)

            if (assets.count > 0) {
                let asset: PHAsset = assets[0];

                let ID: PHImageRequestID = manager.requestImage(
                    for: asset,
                    targetSize: PHImageManagerMaximumSize,
                    contentMode: PHImageContentMode.aspectFill,
                    options: options,
                    resultHandler: {
                        (image: UIImage?, info) in
                        self.messenger.send(onChannel: "multi_image_picker/image/" + identifier + ".original", message: image!.jpegData(compressionQuality: CGFloat(compressionQuality)))
                })

                if(PHInvalidImageRequestID != ID) {
                    return result(true);
                }
            }
            
            return result(FlutterError(code: "ASSET_DOES_NOT_EXIST", message: "The requested image does not exist.", details: nil))
        case "requestMetadata":
            let arguments = call.arguments as! Dictionary<String, AnyObject>
            let identifier = arguments["identifier"] as! String
            let operationQueue = OperationQueue()
            
            let assets: PHFetchResult = PHAsset.fetchAssets(withLocalIdentifiers: [identifier], options: nil)
            operationQueue.addOperation {
                self.readPhotosMetadata(result: assets, operationQueue: operationQueue, callback: result)
            }
            break;
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    func readPhotosMetadata(result: PHFetchResult<PHAsset>, operationQueue: OperationQueue, callback: @escaping FlutterResult) {
        let imageManager = PHImageManager.default()
        result.enumerateObjects({object , index, stop in
            let options = PHImageRequestOptions()
            options.isNetworkAccessAllowed = true
            options.isSynchronous = false
            imageManager.requestImageData(for: object, options: options, resultHandler: { (imageData, dataUTI, orientation, info) in
                operationQueue.addOperation {
                    guard let data = imageData,
                        let metadata = type(of: self).fetchPhotoMetadata(data: data) else {
                            print("metadata not found for \(object)")
                            return
                    }
                    callback(metadata)
                }
            })
        })
    }
    
    static func fetchPhotoMetadata(data: Data) -> [String: Any]? {
        guard let selectedImageSourceRef = CGImageSourceCreateWithData(data as CFData, nil),
            let imagePropertiesDictionary = CGImageSourceCopyPropertiesAtIndex(selectedImageSourceRef, 0, nil) as? [String: Any] else {
                return nil
        }
        return imagePropertiesDictionary
        
    }

    func hexStringToUIColor (hex:String) -> UIColor {
        var cString:String = hex.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()

        if (cString.hasPrefix("#")) {
            cString.remove(at: cString.startIndex)
        }

        if ((cString.count) != 6) {
            return UIColor.gray
        }

        var rgbValue:UInt32 = 0
        Scanner(string: cString).scanHexInt32(&rgbValue)

        return UIColor(
            red: CGFloat((rgbValue & 0xFF0000) >> 16) / 255.0,
            green: CGFloat((rgbValue & 0x00FF00) >> 8) / 255.0,
            blue: CGFloat(rgbValue & 0x0000FF) / 255.0,
            alpha: CGFloat(1.0)
        )
    }
}
